src/OrchardCore.Modules/OrchardCore.Redis/Caching/RedisDistributedCache.cs

```diff
@@ -0,0 +1,219 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Distributed;
+using OrchardCore.Modules;
+
+namespace OrchardCore.Redis.Caching
+{
+    /// <summary>
+    /// Wrapper to ensure the system stays available during Redis outages
+    /// </summary>
+    public class RedisDistributedCache : OrchardCore.Caching.Distributed.IDistributedCache
+    {
+        private readonly IDistributedCache _underlyingCache;
+        private readonly IClock _clock;
+
+        private static DateTime _retryAt;
+        private static bool _redisUnavailable = false;
+        private static object _redisUnavailableLock = new object();
+
+        public RedisDistributedCache(Microsoft.Extensions.Caching.Distributed.IDistributedCache cache, IClock clock)
+        {
+            _underlyingCache = cache;
+            _clock = clock;
```

Comment:
As stated by @Skrypt we don't need to define a new interface

Here we could inject `IOptions<RedisCacheOptions> optionsAccessor` so that we can create an inner `_redisCache` without having to resolve it, just by doing `_redisCache = new RedisCache(optionsAccessor);` https://github.com/dotnet/aspnetcore/blob/db4dc17779dbfbe96b94ad33bc02a368b4a94463/src/Caching/StackExchangeRedis/src/RedisCache.cs#L49

Then, in the redis module startup, not call `AddStackExchangeRedisCache()` that registers the `RedisCache`
https://github.com/dotnet/aspnetcore/blob/db4dc17779dbfbe96b94ad33bc02a368b4a94463/src/Caching/StackExchangeRedis/src/StackExchangeRedisCacheServiceCollectionExtensions.cs#L23-L40

And do what `AddStackExchangeRedisCache()` does unless that we register our wrapper

            services.AddOptions(); // <= not necessary here
            services.Configure(setupAction); // <= idem, so I think only the following line
            services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCacheService>());

Then we would have to implement `IDisposable` so that we can dispose our inner `_redisCache`.

    public class RedisCacheService : IDistributedCache, IDisposable

Not sure yet of the best name ;)

Reply:
Since the code doesn't actually depend on anything Redis specific what if I made it more generic?

The one problem I see with that is I don't have a 'clean room' code implementation of wrapping an existing dependency in services collection (I have an implementation I grabbed from somewhere else so the rights get tricky I think).  

A semi dirty way around that would be to just wrap it in the DocumentManagers constructor in the if (distributedCache == typeof(MemoryCacheDistributedCache) {} block.  I'm not the biggest fan of that though if someone can provide a clean services.AddDecorator(..) extension method then that would be preferred?

Reply:
Ignore the 2nd part of my comment as I re-read your comment and see that you covered that.